tfhe/src/shortint/server_key/bivariate_pbs.rs

```diff
@@ -0,0 +1,534 @@
+    /// let ct2 = cks.encrypt(msg2);
+    /// let modulus = cks.parameters.message_modulus().0 as u64;
+    ///
+    /// // Generate the lookup table for the function f: x, y -> (x * y * x) mod 4
+    /// let acc = sks.generate_lookup_table_bivariate(|x, y| x * y * x % modulus);
+    /// let ct_res = sks.unchecked_apply_lookup_table_bivariate(&ct1, &ct2, &acc);
+    ///
+    /// let dec = cks.decrypt(&ct_res);
+    /// assert_eq!(dec, (msg * msg2 * msg) % modulus);
+    /// ```
+    pub fn unchecked_apply_lookup_table_bivariate(
+        &self,
+        ct_left: &Ciphertext,
+        ct_right: &Ciphertext,
+        acc: &BivariateLookupTableOwned,
+    ) -> Ciphertext {
+        let mut ct_res = ct_left.clone();
+        self.unchecked_apply_lookup_table_bivariate_assign(&mut ct_res, ct_right, acc);
+        ct_res
+    }
+
+    pub fn unchecked_apply_lookup_table_bivariate_assign(
+        &self,
+        ct_left: &mut Ciphertext,
+        ct_right: &Ciphertext,
+        acc: &BivariateLookupTableOwned,
+    ) {
+        let modulus = (ct_right.degree.get() + 1) as u64;
+        assert!(modulus <= acc.ct_right_modulus.0 as u64);
+
+        self.unchecked_scalar_mul_assign(ct_left, acc.ct_right_modulus.0 as u8);
+
+        unchecked_add_assign(ct_left, ct_right);
+
+        // Compute the PBS
+        self.apply_lookup_table_assign(ct_left, &acc.acc);
+    }
+
+    /// Compute a keyswitch and programmable bootstrap.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use tfhe::shortint::gen_keys;
+    /// use tfhe::shortint::parameters::PARAM_MESSAGE_2_CARRY_2_KS_PBS;
+    ///
+    /// // Generate the client key and the server key:
+    /// let (cks, sks) = gen_keys(PARAM_MESSAGE_2_CARRY_2_KS_PBS);
+    ///
+    /// let msg: u64 = 3;
+    /// let msg2: u64 = 2;
+    /// let ct1 = cks.encrypt(msg);
+    /// let ct2 = cks.encrypt(msg2);
+    /// let modulus = cks.parameters.message_modulus().0 as u64;
+    ///
+    /// // Generate the lookup table for the function f: x, y -> (x * y * x) mod 4
+    /// let acc = sks.generate_lookup_table_bivariate(|x, y| x * y * x % modulus);
+    /// let ct_res = sks.apply_lookup_table_bivariate(&ct1, &ct2, &acc);
+    ///
+    /// let dec = cks.decrypt(&ct_res);
+    /// assert_eq!(dec, (msg * msg2 * msg) % modulus);
+    /// ```
+    pub fn apply_lookup_table_bivariate(
+        &self,
+        ct_left: &Ciphertext,
+        ct_right: &Ciphertext,
+        acc: &BivariateLookupTableOwned,
+    ) -> Ciphertext {
+        let ct_left_clean;
+        let ct_right_clean;
+
+        let (ct_left, ct_right) = if self
+            .is_functional_bivariate_pbs_possible(ct_left.noise_degree(), ct_right.noise_degree())
+            .is_err()
+        {
+            // After the message_extract, we'll have ct_left, ct_right in [0, message_modulus[
+            // so the factor has to be message_modulus
+            assert_eq!(ct_right.message_modulus.0, acc.ct_right_modulus.0);
```

Comment:
I think we have a bug here, it's an old one, will open an issue for it, looks like we expect tables to always be shifted by the message modulus but in practice it's not always the best thing to do e.g. noise wise

Reply:
follow up issue https://github.com/zama-ai/tfhe-rs-internal/issues/364